samza-api/src/main/java/org/apache/samza/operators/MessageStream.java

```diff
@@ -47,20 +47,20 @@
    * transformed {@link MessageStream}.
    *
    * @param mapFn the function to transform a message to another message
-   * @param <TM> the type of messages in the transformed {@link MessageStream}
+   * @param <OM> the type of messages in the output {@link MessageStream}
    * @return the transformed {@link MessageStream}
    */
-  <TM> MessageStream<TM> map(MapFunction<M, TM> mapFn);
+  <OM> MessageStream<OM> map(MapFunction<? super M, ? extends OM> mapFn);
 
   /**
    * Applies the provided 1:n function to transform a message in this {@link MessageStream}
    * to n messages in the transformed {@link MessageStream}
    *
    * @param flatMapFn the function to transform a message to zero or more messages
-   * @param <TM> the type of messages in the transformed {@link MessageStream}
+   * @param <OM> the type of messages in the output {@link MessageStream}
    * @return the transformed {@link MessageStream}
    */
-  <TM> MessageStream<TM> flatMap(FlatMapFunction<M, TM> flatMapFn);
+  <OM> MessageStream<OM> flatMap(FlatMapFunction<? super M, ? extends OM> flatMapFn);
```

Comment:
@nickpan47 explained offline. The way to read this is that the user can provide a function that takes a super-type of M as input since it's always safe to cast the actual provided M to its super-type.

E.g., if M == Integer, this signature allows users to provide a more general FilterFunction<Number> instead of a FilterFunction<Integer> if that's what they want to use.